### ISSUE TITLE
Type string is assignable to interface with index signature from number to string

### ISSUE THREAD

**Issue by vostrnad** [microsoft/TypeScript#41228]
**TypeScript Version:**  4.1.0-beta

**Search Terms:** string assignable index signature number 

**Code**

```ts
interface NumberIndexable {
  [index: number]: string;
}

function setValueAtIndex(object: NumberIndexable, index: number, value: string) {
  object[index] = value;
}

setValueAtIndex({}, 0, 'test'); // works fine
setValueAtIndex('oops', 0, 'test'); // runtime error
```

**Expected behavior:**

Second call to setValueAtIndex should raise a compiler error.

**Actual behavior:**

No compiler error is raised, which results in a runtime error because string indeces are read-only.

**Playground Link:** [Playground](https://www.typescriptlang.org/play?ts=4.1.0-beta#code/JYOwLgpgTgZghgYwgAgHIFcC2AjaBJEAEwgA85sAbFAbwChlkBtUYkgLmRC1ygF0OAzmCigA5gG5aAX1q0Y6EAjDAA9iGQCIYAGpwK6CAEEwBVgAoV2AFYQlHDDnxFS5KgBpkLUhy6OoHgDc9A0FhMQBKZDoGSxslZmcSXmQAXmQg-QhJGVpNHWCjE0SzaikPAAYPAHJIISrw8WQAeibkAHcVKABrAWQYUAhcrV1M41NSMyqVFQAHASqK6tqwesaW5CgFZUwUaChO2iA)

**Comment by MartinJohns** [comment]
`string` being assignable to your type is correct and working as intended. A `string` can be accessed with an index signature.

The problem stems from TypeScript not distinguishing between readonly and mutable types, or no way to specify immutability in the type-system. The immutable type `string` should not be assignable to a mutable interface. Issues to track this would be #18770 and #10725.
